Documentation/design/installconfig.md

```diff
@@ -0,0 +1,183 @@
+    IAMRoleName      string            `json:"tectonic_aws_worker_iam_role_name,omitempty" yaml:"iamRoleName,omitempty"`
+    LoadBalancers    []string          `json:"tectonic_aws_worker_load_balancers,omitempty" yaml:"loadBalancers,omitempty"`
+    WorkerRootVolume `json:",inline" yaml:"rootVolume,omitempty"`
+}
+```
+
+#### libvirt specfic configuration
+
+```go
+type Libvirt struct {
+    URI           string `json:"tectonic_libvirt_uri,omitempty" yaml:"uri"`
+    SSHKey        string `json:"tectonic_libvirt_ssh_key,omitempty" yaml:"sshKey"`
+    QCOWImagePath string `json:"tectonic_coreos_qcow_path,omitempty" yaml:"imagePath"`
+    Network       `json:",inline" yaml:"network"`
+    MasterIPs     []string `json:"tectonic_libvirt_master_ips,omitempty" yaml:"masterIPs"`
+}
+
+// Network describes a libvirt network configuration.
+type Network struct {
+    Name      string `json:"tectonic_libvirt_network_name,omitempty" yaml:"name"`
+    IfName    string `json:"tectonic_libvirt_network_if,omitempty" yaml:"ifName"`
+    DNSServer string `json:"tectonic_libvirt_resolver,omitempty" yaml:"dnsServer"`
+    IPRange   string `json:"tectonic_libvirt_ip_range,omitempty" yaml:"ipRange"`
+}
+```
+
+### 4.0 Installer InstallConfig
+
+```go
+type InstallConfig struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata"`
+
+    ClusterID       uuid.UUID `json:"clusterID"`
+    SSHKey          string `json:"sshKey"`
+    Networking      `json:"networking"`
+    Machines        map[string]MachineSet `json:"machines"` // MachineSet https://github.com/kubernetes-sigs/cluster-api/blob/dfa371814c0b39a284662b897ba9b5df4749e973/pkg/apis/cluster/v1alpha1/machineset_types.go#L40
+
+    AWS `json:"aws,omitempty`
+    Libvirt `json:"libvirt,omitempty"`
+}
+
+type Networking struct {
+    Type        NetworkType `json:"type"`
+    MTU         string `json:"mtu"`
+    ServiceCIDR net.IPNet `json:"serviceCIDR"`
+    PodCIDR     net.IPNet `json:"podCIDR"`
+}
+
+// AWS stores all the global configuration that
+// all machinesets use.
+type AWS struct {
+    Region              string `json:"region"`
+    VPCID               string `json:"vpcID"`
+    VPCCIDRBlock        string `json:"vpcCIDRBlock"`
+    AvailabiltyZones    []string `json:"azs"`
+}
+
+// Libvirt stores all the global configuration that
+// all machinesets use.
+type Libvirt struct {
+    URI           string `json:"URI"`
+    QCOWImagePath string `json:"qcowImagePath"`
+    Network       `json:"network,inline"`
+    MasterIPs     []net.IP `json:"masterIPs"`
+}
+
+type LibvirtNetwork struct {
+    Name      string `json:"name"`
+    IfName    string `json:"if"`
+    DNSServer string `json:"resolver"`
+    IPRange   string `json:"ipRange"`
+}
+```
+
+### EC2 MachineSetProvider
+
+```go
+type EC2MachineSetProvider struct {
+    Type string `json:"type"`
+    IAMRoleName string `json:"iamRoleName"`
+    RootVolume string `json:"rootVolume"`
+}
+```
+
+### QEMU MachineSet
```

Comment:
Need to contrast / reference the existing Cluster object in cluster-operator.

Reply:
Or link in with previous work.

Reply:
Added previous work section for cluster-operator.
should the InstallConfig be consistent with cluster-operator's `ClusterDeployment`?
@yifan-gu @crawford @aaronlevy